dol/sources.py

```diff
@@ -108,37 +108,114 @@ def __getitem__(self, k):
     accumulated in a dict that is then returned.
+
+    Args:
+        `*args`: sub-stores used to fan-out the data. These stores will be represented by 
+            their index in the tuple.
+        `default`: default value to return by a store if the requested key is not in the 
+            store. Note that a KeyError is raised if the key is not in any of the stores.
+        `get_existing_values_only`: if True, only return values for stores that contain 
+            the key.
+        `**kwargs`: sub-stores used to fan-out the data, with their own user-defined keys
+
+    Let's define the following sub-stores:
+
+    >>> bytes_store = dict(
+    ...     a=b'a',
+    ...     b=b'b',
+    ...     c=b'c',
+    ... )
+    >>> metadata_store = dict(
+    ...     b=dict(x=2),
+    ...     c=dict(x=3),
+    ...     d=dict(x=4),
+    ... )
+
+    We can create a fan-out reader from these stores:
+
+    >>> reader = FanoutReader(bytes_store, metadata_store)
+    >>> reader['b']
+    {0: b'b', 1: {'x': 2}}
+
+    The reader returns a dict with the values from each store, keyed by the index of the
+    store in the `args` tuple.
+
+    We can also create a fan-out reader passing the stores in kwargs:
+
+    >>> reader = FanoutReader(bytes_store=bytes_store, metadata_store=metadata_store)
+    >>> reader['b']
+    {'bytes_store': b'b', 'metadata_store': {'x': 2}}
+
+    This way, the returned value is keyed by the name of the store.
+
+    We can also mix args and kwargs:
+
+    >>> reader = FanoutReader(bytes_store, metadata_store=metadata_store)
+    >>> reader['b']
+    {0: b'b', 'metadata_store': {'x': 2}}
+
+    Note that the order of the stores is determined by the order of the args and kwargs.
+
+    We can also pass a default value to return if the key is not in the store:
+
+    >>> reader = FanoutReader(
+    ...     bytes_store=bytes_store,
+    ...     metadata_store=metadata_store,
+    ...     default='no value in this store for this key', 
+    ... )
+    >>> reader['a']
+    {'bytes_store': b'a', 'metadata_store': 'no value in this store for this key'}
+
+    If the key is not in any of the stores, a KeyError is raised:
+
+    >>> reader['z']
+    Traceback (most recent call last):
+        ...
+    KeyError: 'z'
+
+    We can also pass `get_existing_values_only=True` to only return values for stores
+    that contain the key:
+
+    >>> reader = FanoutReader(
+    ...     bytes_store=bytes_store,
+    ...     metadata_store=metadata_store,
+    ...     get_existing_values_only=True,
+    ... )
+    >>> reader['a']
+    {'bytes_store': b'a'}
     """
 
-    def __init__(
-        self,
-        stores: Mapping,
-        default=None,
-        # *,
-        keys: Iterable = None,
-        # assert_all_have_key: bool=False,
-    ):
-        self._stores = stores
-        # TODO: More control on what to do with missing keys
+    def __init__(self, *args, default=None, get_existing_values_only=False, **kwargs):
```

Comment:
Note two things with this choice of signature, versus an alternative, say:

```python
def __init__(self, stores: Mapping[Any, Mapping], *, default=None, get_existing_values_only=False): ...
```

First, in your proposal, the user wouldn't be able to name a store `get_existing_values_only` (unlikely) or `default` (a lot more likely). I can't even see a way out of it for them. They don't even have an alternative to make their fanout store, but to rename their store.

Further, the only names they can give their stores through kwargs, are valid python names. Again, if they're working with stores that already have natural names (that might contain spaces, or dashes, etc. they'd be obliged to do some awkward (and possibly buggy) name mapping. 

Further, it's not even true that all stores must be valid python names, because the ones communicated through *args are integers! 

So I'd vote to have the `__init__` be something like the signature I proposed above, but still have a factory `@classmethod` to do what your proposal does, for when it's more convenient (because it **is** more convenient in some cases). By doing so, not only the init is a straightforward assignment (like dataclasses), but you also don't limit yourself.